0000-encryption-api/README.md

```diff
@@ -0,0 +1,135 @@
+- Feature Name: (mal000001, encryption-api)
+- Start Date: 6-Feb-2019
+- RFC PR: (leave this empty)
+- Ursa Issue: (leave this empty)
+- Version: 1
+
+# Summary
+[summary]: #summary
+
+Encryption is considered a primary use for cryptography which Ursa does not
+currently provide. There is demand for providing easy to use encryption/decryption
+interfaces in Ursa. Ursa already has similar interfaces for signing/verification.
+The implemented algorithms provide programmers choices that aim
+to avoid bad parameter interactions. Keys are passed around as data blobs
+allowing consumers to protect them as they see fit.
+
+# Motivation
+[motivation]: #motivation
+
+Hyperledger projects will need to be able to encrypt/decrypt data using
+industry standards. To minimize mistakes and allow other projects good
+encryption schemes, Ursa will provide known good implementations with
+good APIs that are simple enough to avoid confusion.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Ursa should distinguish between two methods for encryption: one for asymmetric algorithms
+like RSA, and one for symmetric like AES. However the interfaces will be very similar.
+
+[Libsodium](https://libsodium.gitbook.io/doc/) for example uses these two types as public-key encryption and secret key encryption.
+
+Public key encryption is generally applied for key exchange or secret key wrapping. It requires two different keys–*public* and *private*, linked together with a one way mathematical
+relationship. The *private* key is used to decrypt a message and cannot be recovered from the *public* key and MUST remain private. The *public* key is used to encrypt a message.
+
+Secret key encryption uses the same key to encrypt and decrypt. It is only shared among all involved parties as a *shared secret key*.
+
+Users will create keys using one of these two types and pass them to the appropriate method.
+Rust typing will prevent a user from passing a secret key to a public key encryption method and visa versa.
+
+Public-Key encryption methods MUST use the naming convention **ALGORITHM or CURVE{SIZE}-SCHEME**. Below are some examples
```

Comment:
Can we generalize this?  Something like "SCHEME--PARAMETERS"?  Curve size doesn't generalize to, say, lattice-based or other assumptions (including the dreaded RSA) nicely.

Reply:
I imagine it looking like this

Ed25519-Chacha20Poly1305
P256-AES128GCM
P521-AES256GCM
Secp256k1-XSalsa20Poly1305

RFCs can be updated when Lattice-based crypto needs to be added. If you want to do it now, what do you recommend for including those schemes?

Reply:
I was hoping for something like

(Function)-(Scheme)-(Parameters)

So this would be like

SymEncrypt-AES-128

Hash functions could also be included in the parameters.  

I think it's a good idea to have the functionality of what is being done in the name.  ED25519 is a curve that you can do whatever you want on it--in theory, you could be describing some kind of encryption scheme with Ed25519-Chacha20Poly1305, although I'm not entirely sure how the hash function would be used.